File: api/src/entities/WorklogSubmission.ts
import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  CreateDateColumn,
  UpdateDateColumn,
  OneToMany,
} from 'typeorm';

import WorklogSubmissionResult from './WorklogSubmissionResult';

export type WorklogTarget = 'TEMPO' | 'JIRA' | 'BOTH';
export type SubmissionStatus = 'PENDING' | 'SUCCESS' | 'FAILED' | 'PARTIAL_SUCCESS';

@Entity('worklog_submission')
export default class WorklogSubmission {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ name: 'request_id', type: 'varchar', unique: true })
  requestId: string;

  @Column({ type: 'varchar', length: 10 })
  target: WorklogTarget;

  @Column({ name: 'tempo_issue_key', type: 'varchar' })
  tempoIssueKey: string;

  @Column({ name: 'external_issue_key', type: 'varchar', nullable: true })
  externalIssueKey: string | null;

  @Column({ name: 'work_date', type: 'varchar', length: 10 })
  workDate: string;

  @Column({ name: 'started_at', type: 'varchar', nullable: true })
  startedAt: string | null;

  @Column({ name: 'time_spent_seconds', type: 'integer' })
  timeSpentSeconds: number;

  @Column({ type: 'text' })
  description: string;

  @Column({ name: 'overall_status', type: 'varchar', length: 20 })
  overallStatus: SubmissionStatus;

  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date;

  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date;

  @OneToMany(() => WorklogSubmissionResult, (result) => result.submission)
  results: WorklogSubmissionResult[];
}

File: packages/planny-core/src/planny_core/models/worklog_submission.py
"""WorklogSubmission model — mirrors ``api/src/entities/WorklogSubmission.ts``.

Table name: ``worklog_submission``
"""

from __future__ import annotations

from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import DateTime, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from planny_core.database import Base

if TYPE_CHECKING:
    from planny_core.models.worklog_submission_result import WorklogSubmissionResult


class WorklogSubmission(Base):
    __tablename__ = "worklog_submission"

    id: Mapped[int] = mapped_column(primary_key=True, autoincrement=True)
    request_id: Mapped[str] = mapped_column(
        String(255),
        unique=True,
    )
    target: Mapped[str] = mapped_column(String(10))
    tempo_issue_key: Mapped[str] = mapped_column(String(255))
    external_issue_key: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )
    work_date: Mapped[str] = mapped_column(String(10))
    started_at: Mapped[str | None] = mapped_column(String(50), nullable=True)
    time_spent_seconds: Mapped[int] = mapped_column(Integer)
    description: Mapped[str] = mapped_column(Text)
    overall_status: Mapped[str] = mapped_column(String(20))

    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
    )

    # ── Relationships ──────────────────────────────────────────────────────
    results: Mapped[list[WorklogSubmissionResult]] = relationship(
        "WorklogSubmissionResult",
        back_populates="submission",
        cascade="all, delete-orphan",
    )

    def __repr__(self) -> str:
        return f"<WorklogSubmission id={self.id} request_id={self.request_id!r}>"
